Replace PyCrypto with tinyaes for encryption support

naufraghi opened the pull request:
[tinyaes](https://github.com/naufraghi/tinyaes-py) is a minimal AES-only library that wraps the C library [tiny-AES-c](https://github.com/kokke/tiny-AES-c).

Currently the library wraps only the CTR mode that can be used in PyInstaller instead of the CFB mode used before.

Reading various sources, CFB is [somewhat similar to CBC](https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Cipher_Feedback_(CFB)) and CTR is suggested as a go-to stream cipher (https://crypto.stackexchange.com/questions/6029/aes-cbc-mode-or-aes-ctr-mode-recommended and links) that can be used instead of CBC.

See the issue comments for more discussion https://github.com/pyinstaller/pyinstaller/issues/2365#issuecomment-578691520

Closes #2365

naufraghi commented on PyInstaller/building/build_main.py:
```diff
@@ -160,6 +160,9 @@ def __init__(self, scripts, pathex=None, binaries=None, datas=None,
                 An optional list of scripts to use as users' runtime hooks. Specified
                 as file names.
+        cipher
+                Ad optional instance of the pyz_crypto.PyiBlockCipher class
```
Typo (will wait for the CI to complete at least once before committing again)
```suggestion
                Add optional instance of the pyz_crypto.PyiBlockCipher class
```

Legorooj replied:
CI build has finished. You can commit this now. Looks good btw.

naufraghi replied:
I'll rebase and push as soon as possibile!

BoboTiG commented:
The new code is much simpler to read, and after so many months there is finally a possible replacement for that critical code part (that I failed to do it myself, twice ^^). :muscle:

naufraghi commented:
Do we need a similar PR for the master too?

naufraghi commented:
Hmm, I'll wait a bit more, I'm adding some test to the library, I'd prefer to point a stable release in `tests/requirements-tools.txt`

BoboTiG commented:
> Do we need a similar PR for the master too?

Nope, target `develop` only.

Legorooj commented:
@naufraghi will review this later today when on PC.

naufraghi commented:
Sorry for the review requests, I was just inspecting the the GitHub gui 🙄

Legorooj commented:
@naufraghi not a problem. I've got to say, this looks good. I'm going to suggest that seeing as you said you don't want the bounty, it goes to the PyInstaller team to fund @htgoebel.

Legorooj commented on tests/functional/test_basic.py:
```diff
@@ -162,17 +162,15 @@ def test_email(pyi_builder):
 
-        # Issue 1663: Crypto feature caused issues when using PyCrypto module.
-        import Crypto.Cipher.AES
+        # An issue similar to the 1663 with PyCrypto may arise for tinyaes
```
I'd remove the comment unless you have any reason to believe that the issue could happen.

naufraghi replied:
I had a look on the issue #1663, and the test is there to check the the module namespace is cleaned up after loading the AES module during the bootstrap.

It seems to be legit, what do you think if I add some more explicit info instead of a plain issue number?

Legorooj replied:
@naufraghi good point on the legit - I reread it and agree now. Just add a bit more info in the comment.

naufraghi replied:
Reworded a bit

Legorooj commented:
Also just to link it: Closes #2365

naufraghi commented:
Added `news/4652.bootloader.rst` and updated both README and `usage.rst`, need to check that everything works as expected (is there some easy way using the CI?).

Legorooj commented on doc/_common_definitions.txt:
```diff
@@ -45,7 +45,7 @@
 .. _plistlib: https://docs.python.org/3/library/plistlib.html
 .. _png2icns: http://icns.sourceforge.net/
-.. _PyCrypto: https://pypi.python.org/pypi/pycrypto/
+.. _`tinyaes`: https://github.com/naufraghi/tinyaes-py
```
reST doesnt need the backticks here

Legorooj commented on README.rst:
```diff
@@ -155,6 +155,6 @@ http://www.pyinstaller.org/funding.html for how to support PyInstaller.
 
 
-.. _PyCrypto: https://www.dlitz.net/software/pycrypto/
+.. _`tinyaes`: https://github.com/naufraghi/tinyaes-py
```
reST doesn't need the backticks here

Legorooj commented on README.rst:
```diff
@@ -64,15 +64,15 @@ PyInstaller is available on PyPI. You can install it through `pip`::
 ------------------------------------
 
-- Python: 
+- Python:
```
No whitespace changes please - that includes all of the other changes you made. The tabs are there for a reason!

naufraghi replied:
I'll revert them all (... but why? They seemed quite inconsistent around the file, that's why I removed them.)

Legorooj replied:
@naufraghi if you wish a detailed explanation ask htgoebel. He's the one who insists on it - and as the project head...

htgoebel replied:
@Legorooj @naufraghi The reason is explained in the manual https://pyinstaller.readthedocs.io/en/stable/development/commit-messages.html#in-detail

Legorooj commented on README.rst:
```diff
@@ -64,15 +64,15 @@ PyInstaller is available on PyPI. You can install it through `pip`::
 
  - Windows XP or newer.
-    
+
```
No whitespace - the tabs are there for a reason

Legorooj commented on README.rst:
```diff
@@ -127,14 +127,14 @@ enhancements on these are welcome.
 bootloader, as we do not ship binary packages. Download PyInstaller
 source, and build the bootloader::
-     
+
```
No whitespace - the tabs are there for a reason

Legorooj commented on README.rst:
```diff
@@ -127,14 +127,14 @@ enhancements on these are welcome.
 
         python setup.py install
-        
+
```
No whitespace - the tabs are there for a reason

naufraghi commented:
@Legorooj did you eventually managed to have a look at the wrapper (and maybe the wrapped library)?

I pinned the PR to `tinyaes ~= 1.0`, and if nothing blocks I'm going to release 1.0.0 later this day.

Legorooj commented:
@naufraghi I skimmed it the other day and plan to give it a more thorough look over it soon. Don't fix the above whitespace until you release 1.0.0 - that'll allow the CI to run without the above mentioned error.

Legorooj commented:
@naufraghi could you also edit the original comment and add `Closes #2365` to the end?

naufraghi commented:
Too late for the whitespaces :/, but I'll wait for the `Closes #2365`

naufraghi commented:
Waiting for https://github.com/naufraghi/tinyaes-py/pull/8 to be merged.

naufraghi commented:
The tinyaes 1.0.0 is on PyPI, the CI here is green again, ready for the final review / merge.

Legorooj commented:
@bjones1 @htgoebel

naufraghi commented:
I use to let the reviewer resolve the discussions, but I can close myself the ones I think I have resolved, let me know the preferred workflow.

bjones1 commented:
Sorry, I lack the time to review this.

Legorooj commented:
@htgoebel Can you please review?

naufraghi commented:
Rebased, just to check everything still works.

Legorooj commented:
@naufraghi I've restarted the failing build. It's running. FYI we're just waiting for @htgoebel to review and merge.

Legorooj commented:
@htgoebel can you look at this PR now? It should be ready to merge, we just need final review.

Legorooj commented:
@htgoebel, can I merge this?

Legorooj commented:
@naufraghi thanks for this feature/bugfix!
